src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketConnection.cs

```diff
@@ -209,7 +216,18 @@ private async Task ProcessReceives()
 
                 input.Advance(bytesReceived);
 
-                var flushTask = input.FlushAsync();
+                var flushTask = input.FlushAsync(isMoreData: !_waitForData);
+                if (!_waitForData)
+                {
+                    if (buffer.Length - bytesReceived >= MinAllocBufferSize)
+                    {
+                        buffer = buffer.Slice(bytesReceived);
```

Comment:
The skipping of the call to GetMemory() doesn't even look like it should be allowed by the PipeWriter. Is this critical for perf?

Reply:
Its what the api would allow